lib/ruby_smb/dcerpc/response.rb

```diff
@@ -16,9 +16,7 @@ class Response < BinData::Record
 
       # PDU Body
       string      :stub, label: 'Stub', read_length: -> { stub_length }
-      string      :auth_pad,
-        onlyif: -> { has_auth_verifier? },
-        length: -> { (16 - (stub.num_bytes % 16)) % 16 }
+      string      :auth_pad, onlyif: -> { has_auth_verifier? }
```

Comment:
This change here and in the request is a bit concerning because it'd alter the behavior. If I understand correctly though from my review, it's necessary because the padding now needs to be calculated by the Kerberos encryptor when GSS authentication is in use. Is that right?

If that's the case, it still seems like there should be a read_length defined for the string field here.

Reply:
So my understanding is that the `read_length` attribute is used in the situation of reading into the structure with the `read` method - i.e. telling it how many bytes to read for that field. In the `Response` class, however, the `read` function has been explicitly overridden to manually fill the `auth_pad` field based on the `sec_trailer.auth_pad_length` field.

Regarding the rationale for removing the `length` field, this was only used (as far as I can tell) in automatically setting the `auth_pad_length` field in the `sec_trailer`; and *that* is the place where behaviour has definitely been changed.

From a high level, it seemed odd to me that the "length" of a bytestring was based not on its actual length, but on the length of a different bytestring. That somehow seems backwards. Certainly it would make sense to have a convenience function (i.e. `how_long_should_the_auth_pad_be`), but automatically setting the "length" of a string based on the length of another string feels non-obvious, as a reader of the code. It also allows the situation where `auth_pad_length` is saying one thing, but the caller can set `auth_pad`'s actual length to something different; so the existing solution isn't a panacea.

In practice, this "magic" made it much harder to reason through the code while debugging a separate issue I was having, and, while it may be strictly correct, it caused all sorts of confounding factors that made the code hard to debug. Once I realised this magic was happening, I was able to make some progress; but as soon as I removed this automated behaviour, and took responsibility as the caller for explicitly calculating and setting the value, I found it so much clearer to read the code, and much easier to reason about what was happening, rather than having values unexpectedly setting themselves implicitly behind my back, and having to add that to my mental model of the code.

This admittedly does put the onus on the caller to ensure the struct has valid padding. But as mentioned earlier, there was already some onus on the caller (to ensure that the `auth_pad` value's actual length matched the automatically-calculated `sec_trailer.auth_pad_length` value); so it's really a trade-off of "code is clearer, but the caller needs to set two values correctly, rather than just one".

To help ensure correctness, I added an assertion in `client.rb`, which will throw an error if the multiple-of-16 requirement is violated. I've just added an extra assertion that `auth_pad_length` is equal to `auth_pad.length` - will push that up in a sec.

So the tl;dr here is: I found the code way more readable once I removed it, it never guaranteed correctness anyway, and we're now asserting on malformed structures, so should catch incorrect uses.

But if we're worried about this being a breaking change for other users of `ruby_smb`, I can revert the change to the automated approach.

Reply:
Okay, I've restored it to the old behaviour, and made that change to make it a bit more readable.